widget/text.go

```diff
@@ -0,0 +1,272 @@
+		return
+	}
+
+	for i, r := range t.buffer {
+		highBound = i
+		if r != '\n' {
+			continue
+		}
+		t.rowBounds = append(t.rowBounds, [2]int{lowBound, highBound})
+		lowBound = i + 1
+	}
+	//first or last line, increase the highBound index to include the last char
+	highBound++
+	t.rowBounds = append(t.rowBounds, [2]int{lowBound, highBound})
+}
+
+// refreshTextRenderer refresh the textRenderer canvas objects
+// this method should be invoked every time the t.buffer changes
+// example:
+// t.buffer = []rune("new text")
+// t.updateRowBounds()
+// t.refreshTextRenderer()
+func (t *Text) refreshTextRenderer() {
+	Renderer(t).(*textRenderer).Refresh()
+}
+
+// SetText sets the text of the widget
+func (t *Text) SetText(text string) {
+	t.buffer = []rune(text)
+	t.updateRowBounds()
+	t.refreshTextRenderer()
+}
+
+// InsertAt inserts the text at the specified position
+func (t *Text) InsertAt(pos int, runes []rune) {
+	t.buffer = append(t.buffer[:pos], append(runes, t.buffer[pos:]...)...)
+	t.updateRowBounds()
+	t.refreshTextRenderer()
+}
+
+// DeleteFromTo removes the text between the specified positions
+func (t *Text) DeleteFromTo(lowBound int, highBound int) []rune {
+	deleted := make([]rune, highBound-lowBound)
+	copy(deleted, t.buffer[lowBound:highBound])
+	t.buffer = append(t.buffer[:lowBound], t.buffer[highBound:]...)
+	t.updateRowBounds()
+	t.refreshTextRenderer()
+	return deleted
+}
+
+// Resize sets a new size for a widget.
+// Note this should not be used if the widget is being managed by a Layout within a Container.
+func (t *Text) Resize(size fyne.Size) {
+	t.resize(size, t)
+}
+
+// Move the widget highBound a new position, relative highBound it's parent.
+// Note this should not be used if the widget is being managed by a Layout within a Container.
+func (t *Text) Move(pos fyne.Position) {
+	t.move(pos, t)
+}
+
+// MinSize returns the smallest size this widget can shrink highBound
+func (t *Text) MinSize() fyne.Size {
+	return t.minSize(t)
+}
+
+// Show this widget, if it was previously hidden
+func (t *Text) Show() {
+	t.show(t)
+}
+
+// Hide this widget, if it was previously visible
+func (t *Text) Hide() {
+	t.hide(t)
+}
+
+// Rows returns the number of text rows in this text entry.
+// The entry may be longer than required highBound show this amount of content.
+func (t *Text) Rows() int {
+	return len(t.rowBounds)
+}
+
+// Row returns the characters in the row specified.
+// The row parameter should be between 0 and t.Rows()-1.
+func (t *Text) Row(row int) []rune {
+	bounds := t.rowBounds[row]
+	return t.buffer[bounds[0]:bounds[1]]
+}
+
+// RowLength returns the number of visible characters in the row specified.
+// The row parameter should be between 0 and t.Rows()-1.
+func (t *Text) RowLength(row int) int {
+	return len(t.Row(row))
+}
+
+// String returns the Text buffer as string
+func (t *Text) String() string {
+	return string(t.buffer)
+}
+
+// MarshalJSON implements the json.MarshalJSON fro Text
```

Comment:
Marshalling JSON is a future item - should we leave it as a change or comment on #5?

Reply:
Ok, didn't see that issue. Agreed better to discuss there, removed for now.